Infragraph LA docs update [IPE-1505] (#2710)
hashicorp/web-unified-docs: content/hcp-docs/content/docs/infragraph/reference/query.mdx

---
page_title: Infragraph query language reference
description: |-

>  The query language allows users to traverse the graph, filter nodes by properties, and navigate edges between resources.

Reviewer: ```suggestion
 The Infragraph query language allows users to traverse the graph, filter nodes by properties, and navigate edges between resources.
```

SEO and all that

                        - [`name`](#property): string | required
                        - [`alias`](#property): string | optional
                - [`comparator`](#comparator): string | required
                - [`right`](#right): map | required
                    - [`value`](#right): map | required unless [`valueList`](#right) defined
                    - [`valueList`](#right) map | required unless [`value`](#right) defined
                        - [`values`](#right) list[[`value`](#right)] | required
            - [`filterGroup`](#filtergroup) map | required unless [`match`](#match) defined
                - [`operator`](#filtergroup) string | required
                - [`filter`](#filtergroup) list[[`match`](#match)] | required 
    - [`edge`](#edge): map | required
        - [`edgeTypes`](#edgetypes): list[string] | required
        - [`direction`](#direction): string | required
        - [`absent`](#absent): boolean | optional
        - [`nodeQuery`](#nodequery): map | required
            - [`node`](#node): map | required
            - [`edge`](#edge): map | optional

## Specification

This section provides details about the fields used to create a `query`.

### query

> A `query` block defines a single Infragraph query. Each `query` block must contain one [`node`](#node) block, and can contain one optional [`edge`](#edge) block.

Reviewer: ```suggestion
A `query` block defines a single Infragraph query. Each `query` block must contain one [`node`](#node) block, and may contain one optional [`edge`](#edge) block.
```

Thinking along the lines of RFC requirement level terms.  https://datatracker.ietf.org/doc/html/rfc2119

- Default: None
- This field is required
- Data type: Object containing one or more fields that define a query
    - One required [`node`](#node)
    - One optional [`edge`](#edge)

### node

> A `node` block defines which nodes to match at the current step of the traversal. You must specify at least one of `unifiedTypes` or `sourceTypes`, and can optionally provide a specific `ids` filter.

Reviewer: ```suggestion
A `node` block defines which nodes to match at the current step of the traversal. You must specify at least one of `unifiedTypes` or `sourceTypes`, and may optionally provide a specific `ids` filter.
```

- Default: None
- This field is required

> - Data type: Object containing at least one [`unifiedTypes`](#unifiedtypes) or [`sourceTypes`](#sourcetypes). You can define both.

Reviewer: ```suggestion
- Data type: Object containing at least one [`unifiedTypes`](#unifiedtypes) or [`sourceTypes`](#sourcetypes). You may define both.
```

    - [`sourceTypes`](#sourcetypes): Matches nodes based on their exact provider-specific type. Does not support wildcards.
    - [`ids`](#ids): Matches specific graph node IDs. Does not support wildcards.
    - [`where`](#where): A list of property filters to apply to the matched nodes.

These dimensions are combined using a logical `AND`. For example, providing both `unifiedTypes=["VIRTUAL_MACHINE"]` and `sourceTypes=["AWS_EC2_INSTANCE"]` looks for nodes that satisfy both constraints.

The following is an example of a `node` block:

```json
{
  "node": {
    "unifiedTypes": ["VIRTUAL_MACHINE"],
    "where": [
      {
        "match": {
          "left": { "property": { "name": "status" } },
          "comparator": "EQUAL",
          "right": { "value": { "string": "running" } }
        }
      }
    ]
  }
}
```

### unifiedTypes

Matches nodes based on their generalized abstract type across providers, such as `["VIRTUAL_MACHINE"]`. Supports a single `"*"` wildcard per query to match any node type. 

- Default: None

> - This field is required unless you define [`sourceTypes`](#sourcetypes). You can define both.

Reviewer: ```suggestion
- This field is required unless you define [`sourceTypes`](#sourcetypes). You may define both.
```

- Data type: List of strings
- Supports a maximum of 5 elements

### sourceTypes

Matches nodes based on their exact provider-specific type, such as `["AWS_EC2_INSTANCE"]`. Does not support wildcards.

- Default: None

> - This field is required unless you define [`unifiedTypes`](#unifiedtypes). You can define both.

Reviewer: ```suggestion
- This field is required unless you define [`unifiedTypes`](#unifiedtypes). You may define both.
```

- Data type: List of strings
- Supports a maximum of 5 elements

### ids

Matches specific graph node IDs. Does not support wildcards.

- Default: None 
- Data type: List of strings
- Supports a maximum of 10 elements

### alias

Optional reference alias to assign to this node. You can reference this alias in [`left`](#left) or [`right`](#right) blocks in a [`where`](#where) clause.

- Default: None
- Data type: string

### where

A list of property filters to apply to the matched nodes. Infragraph matches top-level entries in a `where` block with a logical `AND` operator. Use a `filterGroup` to match multiple filters with an `OR` operator. 

> You can combine multiple `filterGroup` blocks to create a combination of `AND` and `OR` operators. For example, to match the logic of `(A OR B) AND (C OR D)` you can create one `filterGroup` to match at least one `A` or `B` filter, and it must also match one of the `C` or `D` filters.

Reviewer: A worked example of `(A OR B) AND (C OR D)` would be helpful.  Don't _have_ to add it though.

- Default: None
- Data type: A list of [`match`](#match) and [`filterGroup`](#filtergroup) objects
- Supports a maximum of 10 elements

### match

Finds a `node` object that satisfies every `match` clause or the `filterGroup` clause.

- Default: None
- Data type: map
    - [`left`](#left): The property to inspect.
    - [`comparator`](#comparator): The operation to perform.
    - [`right`](#right): The value to compare against the `left` value. 

The following is an example `match` clause using the `IN` comparator:

```json
{
  "match": {
    "left": { "property": { "name": "region" } },
    "comparator": "IN",
    "right": {
      "valueList": {
        "values": [
          { "string": "us-east-1" },
          { "string": "us-west-2" }
        ]
      }
    }
  }
}
```

> ### left

Reviewer: I realize the order of headings is based on the schema above, and I know I'm reading this much more closely (and differently) than a normal reader would.  Just noting that `match` followed by `left` was a bit surprising.

- Default: None
- Data type: string

Refer to [Comparators](#comparators) for more information on the supported comparators.

### right 

The value to compare against the `left` value. 

- Default: None
- Data type: One of `property`, `value`, or `valueList`

Refer to [`property`](#property) for information on the `property` field.

A `value` is an object that contains exactly one data type and one value. The following is a list of example `value` objects:

```json
"right": { "value": { "string": "us-east-1" } }
"right": { "value": { "int": 42 } }
"right": { "value": { "float": 1.5 } }
"right": { "value": { "bool": true } }
"right": { "value": { "timestamp": "2025-02-07T23:00:00-01:00" } }
```

A `valueList` is a list of `value` objects. Infragraph only supports `valueList` objects when you set the `comparator` to `IN` or `NOT_IN`. This block supports a maximum of 10 elements, and all elements must be of the same type.

The following is a an example `right` object that provides a `valueList`:

```json
"right": {
  "valueList": {
    "values": [
      { "string": "us-east-1" },
      { "string": "us-west-2" }
    ]
  }
}
```

### filterGroup

> Defines one `operator` and one or more `filter` used to narrow down the results of the traversal.

Reviewer: ```suggestion
Defines one `operator` and one or more `filter` used to refine the results of the traversal.
```

This one is tough.  "Narrow down" is colloquial.  "Filter" would be the best word.  Reduce, screen, select...don't quite land.

I'm okay if you don't change this.

"Narrow" without "down"?

          "right": { "value": { "string": "production" } }
        }
      },
      {
        "match": {
          "left": { "property": { "name": "is_critical" } },
          "comparator": "EQUAL",
          "right": { "value": { "bool": true } }
        }
      }
    ]
  }
}
```

### edge

An `edge` block dictates how to traverse from the currently matched node to the next set of nodes.

- Default: None
- Data type: Object
    - One required [`edgeTypes`](#edgetypes). Supports a single `*` wildcard per query for any edge type.
    - One required [`direction`](#direction).
    - One optional [`absent`](#absent).
    - One required [`nodeQuery`](#nodequery).

The following is an example of an `edge` block:

```json
{
  "edge": {
    "edgeTypes": ["ATTACHES_TO"],
    "direction": "OUTGOING",
    "nodeQuery": {
      "node": {
        "unifiedTypes": ["ROLE"]
      }
    }
  }
}
```

### edgeTypes

The relationships to traverse, such as `["MANAGES", "CONTAINS"]`. Supports a single `*` wildcard per query for any edge type.

- Default: None
- This field is required
- Data type: List of strings
- Supports a maximum of 5 elements

Infragraph supports the following edge types:

> - `CONTAINS`

Reviewer: Noting these all render in the preview as anchor links

Author: Unfortunately I think this is an issue in DevDot currently where list items in backticks always renders as a link with an empty href

- `RUNS`
- `ATTACHES_TO`
- `APPLIES_TO`
- `PRODUCES`
- `CONNECTS_TO`
- `STORES`
- `HAS_CURRENT`
- `INSTANCE_OF`
- `MANAGES`
- `RUNS_ON`
- `OWNS`
- `SUBSCRIBES_TO`
- `ASSOCIATES_WITH`
- `USES`
- `REPLICATES_TO`
- `REFERS_TO`
- `BOOTS_FROM`

### direction

The direction of the relationship to traverse.

- Default: `OUTGOING`
- This field is required

The following is a list of valid values for this field:

> - `OUTGOING`

Reviewer: Noting these all render in the preview as anchor links too

| `IS_NOT_EMPTY`          | None                           | None                                             | Right operand must be omitted entirely.                                                  |
| `IN`                    | `valueList` (required)        | `string`, `int`, `float` (homogeneous list)      | `bool` and `timestamp` values are rejected inside `valueList`.                          |
| `NOT_IN`                | `valueList` (required)        | `string`, `int`, `float` (homogeneous list)      | Same restrictions as `IN`.                                                               |
| `STARTS_WITH`           | `value` (required)             | `string` only, non-empty                         | Right value must be a non-empty `string` ≤ 1024 chars; any other type/shape is rejected. |

## Limitations

The Query API enforces the following structural limits on queries. If you exceed these limits, Infragraph returns a `400 Bad Request` response containing detailed field violations.

- **Traversal Depth:** Maximum 20 nested `node`/`edge` pairs.
- **Node Filters:**

>   - `unifiedTypes`: Max 5 elements.

Reviewer: "Max" or "Maximum"?

  - `sourceTypes`: Max 5 elements.
  - `ids`: Max 10 elements.
  - Only 1 wildcard (`*`) is allowed across all node types in the entire query.
- **Edge Filters:**
  - `edgeTypes`: Max 5 elements.
  - Only 1 wildcard (`*`) is allowed across all edge types in the entire query.
- **Predicates (`where` clauses):**
  - Max 10 items in a single `where` list.
  - Max 10 items inside a single `filterGroup`.
  - Max predicate nesting depth of 10.
- **Values:**

>   - String values in filters are capped at 1024 characters.

Reviewer: ```suggestion
  - String values in filters cannot exceed 1024 characters.
```
